### ISSUE TITLE
Questions about data process and the model

### ISSUE THREAD

**Issue by Liujingxiu23** [LEEYOONHYUNG/BVAE-TTS#7]
1.Why use min-max-norm

2.Why the following process in the model
   In encoder:
   1）between conv layers:  x = (x1 * torch.sigmoid(x2) + x) / 2 ** 0.5
   2)  final output : value = (key+embedded)/2 ** 0.5

   In BVAE_layer:
   1)in the up process : (inputs+h)/2 ** 0.5
   2)in the down process: outputs = self.post_conv((inputs+h)/2 ** 0.5, mask)

3. Is the process in the model (something/2**0.5) has any relationship with the min-max norm of the mels?

**Comment by LEEYOONHYUNG** [comment]
1. I found that it makes the training stable
2. by dividing by sqrt(2), I can makes the training stable by stabilizing the variance of each layer. It has nothing to do with the normalization.

**Comment by Liujingxiu23** [comment]
@LEEYOONHYUNG Thank you for your reply.

There is another question  I  donot understand. For the BVAE-Layer
<img width="144" alt="bvae-layer" src="https://user-images.githubusercontent.com/49260794/117747061-141c9280-b240-11eb-9bd5-5f27e2cb121a.PNG">
I can understand the posterior distribution  N(delta_mean_k1, delta_sigma_k1) is the down-up path, and the prior distribution N(mean_k, sigma_k) in the top-down path. But what is N(delta_mean_k2, delta_sigma_k2)?  In the top-down path , the only input is h (align_oh * value), what info introduce N(delta_mean_k2, delta_sigma_k2)?

**Comment by LEEYOONHYUNG** [comment]
Thank you for your question. The delta2 parameters are actually different parts from the original NVAE, and as you said, the parameters can actually be considered redundant.

However, we saw that using the additional delta2 parameters really helps us improve the performance.
We speculate that it is because the delta2 can offset the correlation that arises from obtaining delta1 parameters.

If you have another question or your opinion about it, please share it. :D

**Comment by Liujingxiu23** [comment]
@LEEYOONHYUNG  Thank you for your explanation.
There is a question I want to share and discuss with you. As in your paper, you trained 300k interations using the LJSpeech dataset with batch_size=128. I did not do the samething as you.
Instead I trained a multi-speaker model(8 speakers totally 4w sentences) with batch_size=96.
The valid-loss is as the following image.
I wondering why the valid-recon-loss is always decline? Even though I froze learning rate to 0.0001 after iter=400k, the valid-loss is about the same as the following image. Though there is not much imporvement of synthesized wavs after certain iters, processive declined valid-loss is rare for acoustic model . Why this happened? Is the learning rate to small?

**Comment by LEEYOONHYUNG** [comment]
In the case of the LJSpeech, we also see that the losses keep decreasing. However, we train our model for 300k steps because the align and duration losses keep increasing and it can result in unnatural prosody due to incorrect duration modeling.

We speculate that it is because it essentially learns autoencoding rather than converting text to speech.

**Comment by Liujingxiu23** [comment]
@LEEYOONHYUNG  Thank you for your explanation and for your excellent work!
I use hifigan as vocoder and the quality of the synthesized wavs are much better than the wav samples using wavglow in the folder "wav_samples"(I am sorry I can not upload the wavs cause the dataset are private). And the outputs are stable and robust even though alignment are at the time resolution of T_out/4.

**Comment by wizardk** [comment]
> @LEEYOONHYUNG Thank you for your explanation. There is a question I want to share and discuss with you. As in your paper, you trained 300k interations using the LJSpeech dataset with batch_size=128. I did not do the samething as you. Instead I trained a multi-speaker model(8 speakers totally 4w sentences) with batch_size=96. The valid-loss is as the following image. I wondering why the valid-recon-loss is always decline? Even though I froze learning rate to 0.0001 after iter=400k, the valid-loss is about the same as the following image. Though there is not much imporvement of synthesized wavs after certain iters, processive declined valid-loss is rare for acoustic model . Why this happened? Is the learning rate to small?

@Liujingxiu23 Hi, which way did you choose to compute K and V in the multi-speaker model?

**Comment by Liujingxiu23** [comment]
@wizardk  key and value remains as the original code, I just add speaker embeding to value, value+speaker_embeding1 to compute duration, and add another speaker embeding to h, h+speaker_embedding2 to train BVAE blocks.

**Comment by wizardk** [comment]
> @wizardk key and value remains as the original code, I just add speaker embeding to value, value+speaker_embeding1 to compute duration, and add another speaker embeding to h, h+speaker_embedding2 to train BVAE blocks.

@Liujingxiu23 Thanks for your help. Let me reconfirm it by code:

```
def forward(self, text, melspec, text_lengths, mel_lengths, speakers):
        ##### Prepare Mask#####
        self.text_mask, self.mel_mask, self.diag_mask = self.prepare_mask(text_lengths, mel_lengths)
        
        ##### Text #####
        key, value = self.TextEnc(text, self.text_mask)
        
        ##### Bottom_Up #####
        query=self.bottom_up(melspec, self.mel_mask)

        # add for multi-speakers
        speakers_embedded = F.dropout(self.speaker_embedding(speakers), 0.1, training=self.training)
        speakers_embedded = speakers_embedded.unsqueeze(1)
        
        ##### Alignment #####
        h, align = self.get_align(query, key, value, speakers_embedded, text_lengths, mel_lengths, self.text_mask, self.mel_mask)

        ##### Top_Down #####
        mel_pred, kl_loss = self.top_down(h, self.mel_mask)

        ##### Compute Loss #####
        duration_out = self.get_duration(value, speakers_embedded, self.text_mask)
        
        ...

def get_align(self, q, k, v, speakers_embedded, text_lengths, mel_lengths, text_mask, mel_mask):
        ...

        attn_output = (attn_output + speaker_embedded) / 2 ** 0.5
        attn_output = self.Compress(attn_output).transpose(1,2)

        return attn_output, align

def get_duration(self, value, speaker_embedded, mask=None):
        value = (value + speaker_embedded) / 2 ** 0.5
        durations = self.Duration(value.transpose(1,2).detach(), mask)
        return durations
```

Do you mean like this? The most important is adding speaker embeding to h before compress or not?

**Comment by Liujingxiu23** [comment]
> > @wizardk key and value remains as the original code, I just add speaker embeding to value, value+speaker_embeding1 to compute duration, and add another speaker embeding to h, h+speaker_embedding2 to train BVAE blocks.
> 
> @Liujingxiu23 Thanks for your help. Let me reconfirm it by code:
> 
> ```
> def forward(self, text, melspec, text_lengths, mel_lengths, speakers):
>         ##### Prepare Mask#####
>         self.text_mask, self.mel_mask, self.diag_mask = self.prepare_mask(text_lengths, mel_lengths)
>         
>         ##### Text #####
>         key, value = self.TextEnc(text, self.text_mask)
>         
>         ##### Bottom_Up #####
>         query=self.bottom_up(melspec, self.mel_mask)
> 
>         # add for multi-speakers
>         speakers_embedded = F.dropout(self.speaker_embedding(speakers), 0.1, training=self.training)
            ######## I did not use dropout, just nn.Embedding

>         speakers_embedded = speakers_embedded.unsqueeze(1)
>         
>         ##### Alignment #####
>         h, align = self.get_align(query, key, value, speakers_embedded, text_lengths, mel_lengths, self.text_mask, self.mel_mask)
>         ##### I did not add speaker_embeding to get_align, but it may help if added 

>         ##### Top_Down #####
>         mel_pred, kl_loss = self.top_down(h, self.mel_mask)
> 
>         ##### Compute Loss #####
>         duration_out = self.get_duration(value, speakers_embedded, self.text_mask)
>         
>         ...
> 
> def get_align(self, q, k, v, speakers_embedded, text_lengths, mel_lengths, text_mask, mel_mask):
>         ...
> 
>         attn_output = (attn_output + speaker_embedded) / 2 ** 0.5
>         attn_output = self.Compress(attn_output).transpose(1,2)
> 
>         return attn_output, align
> 
> def get_duration(self, value, speaker_embedded, mask=None):
>         value = (value + speaker_embedded) / 2 ** 0.5
           #####  my verison:  durations = self.Duration((value+speaker_embeding).transpose(1,2).detach(), mask)
>         durations = self.Duration(value.transpose(1,2).detach(), mask)
>         return durations
> ```
> 
> Do you mean like this? The most important is adding speaker embeding to h before compress or not?

**Comment by Liujingxiu23** [comment]
@wizardk 
1)speakers_embedded = F.dropout(self.speaker_embedding(speakers), 0.1, training=self.training)
   I did not use dropout here, just nn.Embedding
2)get_align(self, q, k, v, speakers_embedded....)
   I did not use speaker_embediing in this function, but speaker_embedding maybe help here.
3)In get_duration, I just use:
     value = value + speaker_embedded
     durations = self.Duration(value.transpose(1,2).detach(), mask)
4)I also used speaker_embedding in BAVE_Blocks:
   h = h + conv(speaker_embedings) conv here is used to comfort dim match before down or up paths

**Comment by wizardk** [comment]
> @wizardk 1)speakers_embedded = F.dropout(self.speaker_embedding(speakers), 0.1, training=self.training) I did not use dropout here, just nn.Embedding 2)get_align(self, q, k, v, speakers_embedded....) I did not use speaker_embediing in this function, but speaker_embedding maybe help here. 3)In get_duration, I just use: value = value + speaker_embedded durations = self.Duration(value.transpose(1,2).detach(), mask) 4)I also used speaker_embedding in BAVE_Blocks: h = h + conv(speaker_embedings) conv here is used to comfort dim match before down or up paths

@Liujingxiu23 I tried to add speaker embedding only in TextEnc, it works too.
